lib/ActivityStreams.js

```diff
@@ -0,0 +1,144 @@
+/* globals NewTabURL, EventEmitter, XPCOMUtils */
+
+"use strict";
+
+const {Cu} = require("chrome");
+const {data} = require("sdk/self");
+const {PageMod} = require("sdk/page-mod");
+const {PlacesProvider} = require("lib/PlacesProvider");
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource:///modules/NewTabURL.jsm");
+
+XPCOMUtils.defineLazyGetter(this, "EventEmitter", function () {
+  const {EventEmitter} = Cu.import("resource://devtools/shared/event-emitter.js", {});
+  return EventEmitter;
+});
+
+let ActivityStreams = function (options) {
+  console.debug(`ActivityStreams.init`);
+
+  this.options = options ? options : {};
+  this.pageURL = this.options.pageURL ? options.pageURL : data.url("content/activity-streams.html");
+
+  this.workers = new Set();
+  EventEmitter.decorate(this);
+
+  this._setupPageMod();
+  this._setupListeners();
+  NewTabURL.override(this.pageURL);
+};
+
+ActivityStreams.prototype = {
+
+  pageMod: null,
+
+  /**
+   * Send a message to a worker
+   */
+  send(actionName, msg, worker) {
+    worker.port.emit("addon-to-content", {
+      type: actionName,
+      data: msg
+    });
+  },
+
+  /**
+   * Broadcast a message to all workers
+   */
+  broadcast(actionName, msg) {
+    for (let worker of this.workers) {
+      this.send(actionName, msg, worker);
+    }
+  },
+
+  /**
+   * Responds to places requests
+   */
+  _respondToPlacesRequests(msgName, params) {
+    let {msg, worker} = params;
+    PlacesProvider.links.getTopFrecentSites(msg.data).then(links => {
+      this.send("RECEIVE_TOP_FRECENT_SITES", links, worker);
+    });
+  },
+
+
+  /**
+   * Broadcast places changes to pages
+   */
+  _handlePlacesChanges(eventName, data) {
+    this.broadcast("RECEIVE_PLACES_CHANGES", {
+      "event": eventName,
+      "data": data
+    });
+  },
+
+  /**
+   * Sets up various listeners for the pages
+   */
+  _setupListeners() {
+    PlacesProvider.links.on("deleteURI", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("clearHistory", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("linkChanged", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("manyLinksChanged", this._handlePlacesChanges.bind(this));
+
+    this.on("REQUEST_TOP_FRECENT_SITES", this._respondToPlacesRequests.bind(this));
+  },
+
+  /**
+   * Turns off various listeners for the pages
+   */
+  _removeListeners() {
+    PlacesProvider.links.off("deleteURI", this._handlePlacesChanges);
+    PlacesProvider.links.off("clearHistory", this._handlePlacesChanges);
+    PlacesProvider.links.off("linkChanged", this._handlePlacesChanges);
+    PlacesProvider.links.off("manyLinksChanged", this._handlePlacesChanges);
+
+    this.off("REQUEST_TOP_FRECENT_SITES", this._respondToPlacesRequests);
+  },
+
+  /**
+   * Sets up communications with the pages
+   */
+  _setupPageMod() {
```

Comment:
I vetoed on this one. We don't need custom pagemod creation yet. We can take a look in round 2 in case that breaks.

make `setupPageMod` into `createPageMod` doesn't work unless we pass the paramers the function needs, so the function signature would look like `_createPageMod(options.pageURL, this.workers, this)`, which make this a PageMod factory, where param 1 is the url to attach to, this.workers becomes the `Set` to manage workers in and `this` is the EventEmitter to use.

Seemed a bit overkill for now.